### PR TITLE
fix(cardano-services): wrong tx onput `txId` value

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context
The `txId` for the inputs is different than the one provided by Blockfrost.
Cardano services have the tx id of the current transaction, Blockfrost has the tx id of the tx that had that input as output

Requests to both `cardano-services` and `blockfrost` in preview:
image (1)

image (2)

# Proposed Solution

Extend DB query to extract the correct tx input id which represents the transaction(source) id where the input was spent as an output. We can conclude that the tx input `txId` is equal to the source transaction `id`.

We extend the query because before the final mapping to `Cardano.TxIn` return type, we need both `tx_input_id` and `tx_source_id` to filter and map by.

# Important Changes Introduced
- extract the correct `tx_source_id` from db and use it as input `txId`
- update the mappings and types
- added a new e2e test

Introduced e2e test - [txChainHistory.test.ts](https://github.com/input-output-hk/cardano-js-sdk/pull/460/files#diff-6de00ab9b73e5cc2dabcfa65c3f90552a9b8bd4ba27047d7c05e029084f5ca44) which verifies that signed transaction and the same transaction found from the history, must have the same body.

The e2e test is failing without the introduced fix, after the fix - it passes. CI build: [CI build: Test e2e step](https://github.com/input-output-hk/cardano-js-sdk/actions/runs/3128775424/jobs/5077038709)

# Notes
Due to missing enough DB test data, we observe an empty `tx.body.inputs:[]` array in the snapshots. We might need to extend our db-fixtures with more txs/blocks.

Based on the updated snapshot test [Here](https://github.com/input-output-hk/cardano-js-sdk/compare/fix/wrong-tx-id-for-tx-inputs?expand=1#diff-044edd9417ac4bfbdcfdcedd13705ade39a95cac336e0372b3b1b5544cd520bdR791), we can state that the introduced changes fix the wrong tx input `txId` value - there is no duplication between source tx id and current tx input id.

### CODE REVIEW

File: packages/cardano-services/test/ChainHistory/__snapshots__/ChainHistoryHttpService.test.ts.snap [hunk 1]
```diff
@@ -1486,7 +788,7 @@ Array [
         Object {
           "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
           "index": 0,
-          "txId": "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d",
+          "txId": "face165bd7aa8d0d661cf1ceaa4e35d7611be3b1c7997da378c547aa2464a4fd",
```
Comment (IvayloAndonov) [inline packages/cardano-services/test/ChainHistory/__snapshots__/ChainHistoryHttpService.test.ts.snap:791]: Updated input `txId` value

File: packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  it('submit a transaction and find it in chain history', async () => {
+    const tAdaToSend = 10_000_000n;
+    const saWallet = wallet.wallet;
+
+    await firstValueFrom(saWallet.syncStatus.isSettled$.pipe(filter((isSettled) => isSettled)));
+
+    const [{ address: sendingAddress }] = await firstValueFrom(saWallet.addresses$);
+    const [{ address: receivingAddress }] = await firstValueFrom(saWallet.addresses$);
```
Comment (mkazlauskas) [inline packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts:30]:
```suggestion
    const receivingAddress = sendingAddress;
```

File: packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import { getWallet } from '../../../src';
+import { logger } from '@cardano-sdk/util-dev';
+
+export type TestWallet = Awaited<ReturnType<typeof getWallet>>;
+
+describe('tx chain history', () => {
+  let wallet: TestWallet;
+
+  beforeEach(async () => {
+    wallet = await getWallet({ env, logger, name: 'Sending Wallet', polling: { interval: 50 } });
```
Comment (mkazlauskas) [inline packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts:16]: Other props don't seem to be used
```suggestion
    ({ wallet }) = await getWallet({ env, logger, name: 'Sending Wallet', polling: { interval: 50 } });
```

File: packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    const signedTx = await unsignedTx.sign();
+    await signedTx.submit();
+
+    logger.info(
+      `Submitted transaction id: ${signedTx.tx.id}, inputs: ${signedTx.tx.body.inputs} and outputs:${signedTx.tx.body.outputs}.`
+    );
+
+    // Wait until wallet two is aware of the funds.
+    await firstValueFrom(saWallet.balance.utxo.total$.pipe(filter(({ coins }) => coins >= tAdaToSend)));
```
Comment (mkazlauskas) [inline packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts:51]: This seems to be redundant as you're awaiting for tx and not asserting the balance
```suggestion
```

File: packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    // Assert
+    expect(txFoundInHistory).toBeDefined();
+    expect(txFoundInHistory.id).toEqual(signedTx.tx.id);
+    expect(txFoundInHistory.body.inputs.length).toEqual(signedTx.tx.body.inputs.length);
+    expect(txFoundInHistory.body.outputs.length).toEqual(signedTx.tx.body.outputs.length);
+    expect(txFoundInHistory.body.inputs.every((txIn, i) => deepEquals(signedTx.tx.body.inputs[i], txIn))).toBe(true);
+    expect(
+      txFoundInHistory.body.outputs.every((txOut, i) => signedTx.tx.body.outputs[i].value.coins === txOut.value.coins)
+    ).toBe(true);
```
Comment (mkazlauskas) [inline packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts:74]: This would cover other body properties too
```suggestion
    expect(txFoundInHistory.body).toEqual(signedTx.tx.body);
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Just a few suggested improvements for the e2e test

**Comment by rhyslbw** [comment]
@IvayloAndonov I'm finalising the change in #463 for you to use instead, which will cover the missing snapshot data

File: packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts (hunk 5 above)
Comment (IvayloAndonov) [reply packages/e2e/test/wallet/SingleAddressWallet/txChainHistory.test.ts]: All the e2e test feedback is applied under 1bbf6d29661f08f7030a02c3316ae7d54b3d4ff3 @mkazlauskas

File: packages/e2e/test/wallet/util.ts [hunk 6]
```diff
@@ -43,3 +44,8 @@ export const walletReady = (wallet: ObservableWallet) =>
     'Took too long to be ready',
     SYNC_TIMEOUT
   );
+
+export const normalizeTxBody = (body: any) => {
```
Comment (mkazlauskas) [inline packages/e2e/test/wallet/util.ts:48]:
```suggestion
export const normalizeTxBody = (body: TxBodyAlonzo | NewTxBodyAlonzo) => {
```

**Review by mkazlauskas — DISMISSED** [review]
Nice work! No need for a separate commit for the 1 requested change

**Review by mkazlauskas — APPROVED** [review]

**Review by rhyslbw — DISMISSED** [review]
Looks good @IvayloAndonov. Let's ignore the snapshot data issues, as the e2e test covers the scenario, and this fix is now getting urgent to resolve. I'll update the integration tests when I replace the testnet snapshot. Please squash

**Review by rhyslbw — APPROVED** [review]

**Review by mkazlauskas — DISMISSED** [review]

**Review by rhyslbw — APPROVED** [review]
